pkg/controller/migdirect/rsync.go

```diff
@@ -0,0 +1,827 @@
+package migdirect
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"golang.org/x/crypto/ssh"
+	"gopkg.in/yaml.v2"
+	"text/template"
+	//  liberr "github.com/konveyor/controller/pkg/error"
+	"github.com/konveyor/mig-controller/pkg/compat"
+	routev1 "github.com/openshift/api/route/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+	//"k8s.io/apimachinery/pkg/types"
+)
+
+type pvc struct {
+	Name string
+}
+
+type rsyncConfig struct {
+	SshUser   string
+	Namespace string
+	Password  string
+	PVCList   []pvc
+}
+
+// TODO: Parameterize this more to support custom
+// user/pass/networking configs from migdirect spec
+const rsyncConfigTemplate = `apiVersion: v1
+kind: ConfigMap
+metadata:
+  labels:
+    purpose: rsync
+data:
+  rsyncd.conf: |
+    syslog facility = local7
+    read only = no
+    list = yes
+    max = 3
+    auth users = {{ .SshUser }}
+    secrets file = /etc/rsyncd.secrets
+    hosts allow = ::1, 127.0.0.1, localhost
+    uid = root
+    gid = root
+    {{ range $i, $pvc := .PVCList }}
+    [{{ $pvc.Name }}]
+        comment = archive for {{ $pvc.Name }}
+        path = /mnt/{{ $.Namespace }}/{{ $pvc.Name }}
+        uid = root
+        gid = root
+        list = yes
+        hosts allow = ::1, 127.0.0.1, localhost
+        auth users = {{ $.SshUser }}
+        secrets file = /etc/rsyncd.secrets
+        read only = false
+   {{ end }}
+`
+
+func (t *Task) areRsyncTransferPodsRunning() (bool, error) {
+	// Get client for destination
+	destClient, err := t.getDestinationClient()
+	if err != nil {
+		return false, err
+	}
+
+	pvcMap := t.getPVCNamespaceMap()
+
+	selector := labels.SelectorFromSet(map[string]string{
+		"app":     "migdirect-rsync-transfer",
+		"owner":   "migdirect",
+		"purpose": "rsync",
+	})
+	for ns, _ := range pvcMap {
+		pods := corev1.PodList{}
+		err = destClient.List(
+			context.TODO(),
+			&k8sclient.ListOptions{
+				Namespace:     ns,
+				LabelSelector: selector,
+			},
+			&pods)
+		if err != nil {
+			return false, err
+		}
+		for _, pod := range pods.Items {
+			if pod.Status.Phase != corev1.PodRunning {
+				return false, nil
+			}
+		}
+	}
+
+	return true, nil
+
+	// Create rsync transfer pod on destination
+
+	// Create rsync client pod on source
+}
+
+// Generate SSH keys to be used
+// TODO: Need to determine if this has already been generated and
+// not to regenerate
+func (t *Task) generateSSHKeys() error {
```

Comment:
Seems like a good thing here could be to generate these and put them in a secret correlated to the MigDirect. Then if the secret already exists just read from there.

Reply:
Thanks I agree. I will add as a TODO.